src/repr/type.ml

```diff
@@ -371,3 +371,60 @@ module Unboxed = struct
 
   let size_of = Type_size.unboxed
 end
+
+let ref : type a. a t -> a ref t = fun a -> map a ref (fun t -> !t)
+let lazy_t : type a. a t -> a Lazy.t t = fun a -> map a Lazy.from_val Lazy.force
+
+let seq : type a. a t -> a Seq.t t =
+ fun a ->
+  let elt_equal = unstage @@ equal a in
+  let elt_compare = unstage @@ compare a in
+  let rec compare (s1 : a Seq.t) (s2 : a Seq.t) =
+    match (s1 (), s2 ()) with
+    | Nil, Nil -> 0
+    | Cons _, Nil -> 1
+    | Nil, Cons _ -> -1
+    | Cons (x, xf), Cons (y, yf) ->
+        let ord = elt_compare x y in
+        if ord <> 0 then ord else compare xf yf
+  in
+  let rec equal (s1 : a Seq.t) (s2 : a Seq.t) =
+    match (s1 (), s2 ()) with
+    | Nil, Nil -> true
+    | Cons _, Nil | Nil, Cons _ -> false
+    | Cons (x, xf), Cons (y, yf) -> elt_equal x y && equal xf yf
+  in
+  map ~compare:(stage compare) ~equal:(stage equal) (list a) List.to_seq
+    List.of_seq
+
+let stack : type a. a t -> a Stack.t t =
+ fun a -> map (seq a) Stack.of_seq Stack.to_seq
```

Comment:
You might be reversing the stack here.

```ml
let () =
  let l = [42; 8; 9; 6; 4] |> List.to_seq |> Stack.of_seq in
  Printf.eprintf "\n%!";
  Stack.iter (Printf.eprintf "%d\n%!") l;
  Printf.eprintf "\n%!";
  let l = l |> Stack.to_seq |> Stack.of_seq in
  Stack.iter (Printf.eprintf "%d\n%!") l;
```
prints
```
4
6
9
8
42

42
8
9
6
4
```

Reply:
Good catch. I'll add tests for that.